treecorr/binnedcorr2.py

```diff
@@ -1300,17 +1303,54 @@ def _make_cov_design_matrix(corrs, plist, func, name):
     # the original.
     corrs = [c.copy() for c in corrs]
 
+    # We can't pickle functions to send via MPI, so have to do this here.
+    if func is None:
+        func = lambda corrs: np.concatenate([c.getStat() for c in corrs])
+
+    # We also can't send generators using MPI, so instead we need
+    # to create them here.
```

Comment:
So during the CO call just now, I was playing around with an iterators implementation that would be picklable.  Here is what I came up with.
```
    class JK_Pair_Iterator(collections.Iterator):
        def __init__(self, res, np1, np2):
            self.res = res
            self.np1 = np1
            self.np2 = np2

        def with_i(self, i):
            ret = copy.copy(self)
            ret.i = i
            return ret

        def __iter__(self):
            if self.np2 == 1:
                # k=0 here
                gen = ((j,k) for j,k in self.res.keys() if j!=self.i)
            elif self.np1 == 1:
                # j=0 here
                gen = ((j,k) for j,k in self.res.keys() if k!=self.i)
            else:
                # For each i:
                #    Select all pairs where neither is i.
                assert self.np1 == self.np2
                gen = ((j,k) for j,k in self.res.keys() if j!=self.i and k!=self.i)
            self.gen = iter(gen)
            return self

        def __next__(self):
            return next(self.gen)

    def _jackknife_pairs(self):
        it = self.JK_Pair_Iterator(self.results, self.npatch1, self.npatch2)
        return [it.with_i(i) for i in range(self.npatch1)]
```
It's a bit more boilerplate than the old (on current main) version, but it's not too terrible.  And it keeps all the jackknife pairing logic separate from the code for building the design matrix, which I would prefer.

I'm not precisely sure whether this approach will end up pickling the results dict n_patch times, which might be prohibitively slow (or memory demanding).  But if so, we could leave that out of the initial construction, and add a method `assign_results(self, results)` which could be called on the other side of the pickle to each generator, so only one results dict would go through the pickling process.

Anyway, if you want to try fleshing this approach out for the other 4 to make sure it all works, that would be great.  Or if you don't want to, that's totally fine. I can implement it after the merge.

Reply:
I did a little check on the length of the pickle strings, and I think pickle is smart enough to notice that the various results dicts are all identical and only pickle that once.  Pickling the whole plist array isn't much longer than pickling just one of the elements:
```
len(plist) =  16
len pickle(plist) =  57127
len pickle(plist[0]) =  56568
```
So probably this will be fine as is.